fact.go

```diff
@@ -0,0 +1,114 @@
+package xbrl
+
+import (
+	"encoding/xml"
+	"errors"
+	"strconv"
+)
+
+// ErrInvalidNumericFact is returned when a numeric fact does not conform to the XBRL spec and a numeric value cannot be extracted from it.
+// For example, if it's a fraction type fact and the denominator is 0, or the fact is malformed and missing attributes.
+var ErrInvalidNumericFact = errors.New("numeric fact contains invalid data")
+
+// NilFact is a fact in an XBRL document that has an xsi:nil set to a truthy value.
+// These facts still must have a contextRef, but nothing else is guaranteed.
+//
+// Detailed info on Facts in general can be found here:
+// https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.6
+type NilFact struct {
```

Comment:
I'm not clear on how you'd know which of these fact types to use when unmarshalling. Shouldn't we have a single fact type that unmarshals any kind of fact and then determine it's kind after it is parsed?

Reply:
The XBRL custom unmarshaller would know what type of fact it is based on certain attributes (nil facts have `xsi:nil` attributes, numeric facts have `unitRef` attributes, everything else is a non-numeric fact).

I was sort of on the fence about doing it the way you mention and the way I ended up with. I liked the idea of smaller structs for particular types of facts because it would make using this library a bit easier, although it does involve a slightly more complicated unmarshaller to get right. 

Plus a more generic `Fact` that would have a bunch of nilable properties could be awkward to work with.

I'm open to going the route you mention though if you think it makes more sense, particularly because that's the way the `Period` type works in this same repo as well.

The generalized `Fact` struct would essentially be the same as the current `NumericFact` along with a `Nil` attribute to capture `NilFacts` (plus probably some more comments describing what properties you can expect to be non-nil depending on the type of the Fact)

Reply:
I like the idea of it being consistent with the Period type. The nil-able fields is unfortunate, but maybe we can provide a way to convert a general Fact to a specific one after the parser steps. Something like `if fact.Type() == NilFactType { nf := fact.NilFact() }`?

Reply:
Changed it to use a single model. Instead of specific fact structs, I added an `IsValid()` function so you can be confident the fields you expect to exist do exist (depending on the type of fact). LMK what you think